media-source-respec.html

```diff
@@ -829,19 +829,17 @@
           <p>Follow these steps when <a def-id="duration"></a> needs to change to a <var>new duration</var>.</p>
           <ol>
             <li>If the current value of <a def-id="duration"></a> is equal to <var>new duration</var>, then return.</li>
-            <li>Set <var>old duration</var> to the current value of <a def-id="duration"></a>.
+			<li>If <var>new duration</var> is less than the highest start time reported by the <a def-id="buffered"></a> attribute across all <a>SourceBuffer</a> 
+			objects in <a def-id="sourceBuffers"></a>, then throw an <a def-id="invalid-state-error"></a> exception and abort these steps.</li>
+			  <p class="note">Duration reductions must first use <a def-id="remove"></a> to reduce the buffered range before updating <a def-id="duration"></a>.</p>
+            <li>Set <var>old duration</var> to the current value of <a def-id="duration"></a>.</li>
             <li>Update <a def-id="duration"></a> to <var>new duration</var>.</li>
-            <li>If the <var>new duration</var> is less than <var>old duration</var>, then run the
-              <a def-id="range-removal"></a> algorithm with <var>new duration</var> and
-              <var>old duration</var> as the start and end of the removal range.
-
-              <p class="note">This preserves audio frames and text cues that start before and end after the <a def-id="duration"></a>.</p></li>
-            <li>If a user agent is unable to partially render audio frames or text cues that start before and end after the <a def-id="duration"></a>, then run the following steps:
+            <li>If a user agent is unable to partially render audio frames or text cues that start before and end after the <a def-id="duration"></a>, then run the following steps:</li>
+			  <p class="note">This condition can occur because the <a def-id="coded-frame-removal-algorithm"></a> preserves audio frames and text cues that start before and end after the <a def-id="duration"></a>.</p>
```

Comment:
> You seem to be proposing changes to the error criteria for duration reductions that don't first do a remove(). I believe to do that, we need to make a matching change to the Coded Frame Removal Algorithm so that remove() results will pass the check.

Changing the Coded Frame Removal Algorithm (CFRA) to comply with "highest buffered end time" check on subsequent duration reduction would probably regress some important playback scenarios: consider a long video frame or text cue that begin before the eventual new duration and end after it: we probably wouldn't want CFRA to remove those.

> Separate from that, using latest start time as the check aligns with the current remove() algorithm. Is there a problem with it currently that suggests broader change is needed?

I think this *might* be the route forward, however we need a clear definition in the spec of "the latest start time"; it's very rarely going to be the highest range start time in any SourceBuffer.buffered. And it's below the app-visible highest buffered end time. And it will still unfortunately require we "bump" the duration higher (following the case for platforms that can't handle partial frames/cues). I laid out a larger analysis that I can share, but it boiled down to: we need a precise definition for "highest start time", which I believe needs to be something like "highest buffered presentation timestamp across all SourceBuffers' track buffers" (HBPT, below). Ideally, this would be a MediaSource attribute exposed to a web app, so that it could make a reasonable decision in the following example cases where the app wants to change duration to X, where:
* X < HBPT: app should expect exception
* X >= HBPT and <= highest buffered end time across all SourceBuffers' track buffers' track ranges: app shouldn't expect exception, but actual resulting duration might instead be >X if the platform couldn't support partial audio frames/text cues.
* Otherwise, app should expect no exception or special casing as a result of spec changes here.